pkg/plugins/middlewares.go

```diff
@@ -34,30 +39,69 @@ func (b Builder) Build(pName string, config map[string]interface{}, middlewareNa
 type middlewareBuilder struct {
 	fnNew          reflect.Value
 	fnCreateConfig reflect.Value
+	pluginType     pluginType
+	wasmPath       string
 }
 
-func newMiddlewareBuilder(i *interp.Interpreter, basePkg, imp string) (*middlewareBuilder, error) {
+func newWasmMiddlewareBuilder(wasmPath string) *middlewareBuilder {
+	return &middlewareBuilder{
+		wasmPath:   wasmPath,
+		pluginType: PluginTypeWasm,
+	}
+}
+
+func newYaegiMiddlewareBuilder(i *interp.Interpreter, basePkg, imp string) (*middlewareBuilder, error) {
+	builder := &middlewareBuilder{
+		pluginType: PluginTypeYaegi,
+	}
+
 	if basePkg == "" {
 		basePkg = strings.ReplaceAll(path.Base(imp), "-", "_")
 	}
+	var err error
 
-	fnNew, err := i.Eval(basePkg + `.New`)
+	builder.fnNew, err = i.Eval(basePkg + `.New`)
 	if err != nil {
 		return nil, fmt.Errorf("failed to eval New: %w", err)
 	}
 
-	fnCreateConfig, err := i.Eval(basePkg + `.CreateConfig`)
+	builder.fnCreateConfig, err = i.Eval(basePkg + `.CreateConfig`)
 	if err != nil {
 		return nil, fmt.Errorf("failed to eval CreateConfig: %w", err)
 	}
+	return builder, nil
+}
 
-	return &middlewareBuilder{
-		fnNew:          fnNew,
-		fnCreateConfig: fnCreateConfig,
-	}, nil
+func (p middlewareBuilder) newWasmHandler(ctx context.Context, next http.Handler, cfg reflect.Value, middlewareName string) (http.Handler, error) {
```

Comment:
Instead of having one `middlewareBuilder` and two methods named `newWasmHandler` and `newYaegiHandler`, you should have an interface and two implementations ( one for wasm and one for yaegi )

Reply:
its bigger change :)